### DISCUSSION TITLE
`DrawImage` overload with `BitmapInterpolationMode` parameter

### DISCUSSION THREAD

**Discussion by arklumpus** [AvaloniaUI/Avalonia#12616] [Q&A] [answered]
Hi! I'm in the process of updating one of my projects to Avalonia 11, and I noticed that up to commit 4782e1b by @Gillibald, the `DrawingContext.DrawImage` method had overloads with an optional `BitmapInterpolationMode` parameter. This parameter, however, has been removed by that commit.

Is there a way to set a drawing operation-specific bitmap interpolation mode in Avalonia 11, so that one can draw images with different interpolation modes on the same `DrawingContext`? The effect I would like to obtain is this:

![image](https://github.com/AvaloniaUI/Avalonia/assets/32712606/e026908d-2305-4357-8130-ec0ee189e094)

Where the left part of the image is rendered using `BitmapInterpolationMode.None` (which used to be called `Default`) and the right part of the image is rendered using `BitmapInterpolationMode.HighQuality`.

I have tried to call `RenderOptions.SetBitmapInterpolationMode` just before the call to `DrawImage`, but unsurprisingly it doesn't work (all images are rendered using the same interpolation mode, i.e. the one that has been set last).

**Comment by maxkatz6** [comment]
The same PR adds a replacement.
Old render parameters were moved to "RenderOptions.***" attached properties. See xaml example in the description https://github.com/AvaloniaUI/Avalonia/pull/9584

**Reply by arklumpus** [reply]
Thanks, I saw that, but my understanding is that `RenderOptions` apply to the whole element, and not to individual draw operations.

I'm trying to do something like this:

```
public override void Render(DrawingContext context)
{
    context.DrawImage(image1, sourceRect1, destRect1, BitmapInterpolationMode.Default); // Now it's called None
    
    context.DrawImage(image2, sourceRect2, destRect2, BitmapInterpolationMode.HighQuality);
}
```

**Reply by timunie** [reply]
```cs
public override void Render(DrawingContext context)
{
    RenderOptions.SetBitmapInterpolationMode(this, BitmapInterpolationMode.None);
    context.DrawImage();
    RenderOptions.SetBitmapInterpolationMode(this, BitmapInterpolationMode.HighQuality);
    context.DrawImage();
}
```
can you try this?

**Reply by arklumpus** [reply]
I just tried, but as expected calling `RenderOptions.SetBitmapInterpolationMode` in the `Render` method does nothing until the next render pass.

**Reply by timunie** [reply]
@Gillibald any idea?

**Reply by Gillibald** [reply]
You need to use multiple visuals to achieve this because RenderOptions are applied per visual.

**Reply by arklumpus** [reply]
Sadly, that's as I feared, thank you for confirming!

**Reply by timunie** [reply]
that was last week 😆 ... This week we have #12734 which should solve your issue. If you want to try it out, your feedback is welcome.

**Reply by arklumpus** [reply]
Great, thank you! Yes, this does solve the issue (I had to upgrade the language version because init-only setters are only supported from C# 9.0, but that shouldn't be a problem).

**Comment by Gillibald** [comment, marked as answer]
https://github.com/AvaloniaUI/Avalonia/pull/12734
